app/javascript/lib/validation/expression-matrices-validation.js

```diff
@@ -0,0 +1,265 @@
+ */
+function validateColumnNumber(line, isLastLine, headers, lineNum, dataObj) {
+  const issues = []
+  dataObj.rowsWithIncorrectColumnNumbers = dataObj.rowsWithIncorrectColumnNumbers ? dataObj.rowsWithIncorrectColumnNumbers : []
+  // use the first header row to determine the correct number of columns all rows should have
+  const correctNumberOfColumns = headers[0].length
+
+  if (correctNumberOfColumns !== line.length) {
+    dataObj.rowsWithIncorrectColumnNumbers.push(lineNum)
+  }
+
+  const numBadRows = dataObj.rowsWithIncorrectColumnNumbers.length
+  if (isLastLine && numBadRows > 0) {
+    const rowText = numBadRows > 1 ? 'rows' : 'row'
+    const maxLinesToReport = 10
+    let notedBadRows = dataObj.rowsWithIncorrectColumnNumbers.slice(0, maxLinesToReport).join(', ')
+    if (numBadRows - maxLinesToReport > 0) {
+      notedBadRows += ` and ${numBadRows - maxLinesToReport} more rows`
+    }
+
+    const msg = `All rows must have the same number of columns. ` +
+      `Please ensure the number of columns for ${rowText}: ${notedBadRows}, ` +
+      `matches the file-header-specified number of ${correctNumberOfColumns} columns-per-row.`
+    issues.push(['error', 'format:mismatch-column-number', msg])
+  }
+
+  return issues
+}
+
+/**
+ * Verify sparse matrix has no blank lines
+ */
+function validateSparseNoBlankLines(line, isLastLine, lineNum, dataObj) {
+  const issues = []
+  dataObj.blankLineRows = dataObj.blankLineRows ? dataObj.blankLineRows : []
+
+  // if the line is empty, note it
+  if (!line.trim()) {
+    dataObj.blankLineRows.push(lineNum)
+  }
+
+  const numBadRows = dataObj.blankLineRows.length
+  if (isLastLine && numBadRows > 0) {
+    const rowText = numBadRows > 1 ? 'rows' : 'row'
+    const maxLinesToReport = 10
+    let notedBadRows = dataObj.blankLineRows.slice(0, maxLinesToReport).join(', ')
+    if (numBadRows - maxLinesToReport > 0) {
+      notedBadRows += ` and ${numBadRows - maxLinesToReport} more rows`
+    }
+
+    const msg = `Please ensure there are no blank rows in the file. ` +
+      `Remove or replace the following ${rowText}: ${notedBadRows}, `
+    issues.push(['error', 'format:empty-row', msg])
+  }
+
+  return issues
+}
+
+/**
+ * Verify sparse matrix column numbers match
+ */
+function validateSparseColumnNumber(line, isLastLine, lineNum, dataObj) {
+  const issues = []
+  dataObj.rowsWithWrongColumnNumbers = dataObj.rowsWithWrongColumnNumbers ? dataObj.rowsWithWrongColumnNumbers : []
+  dataObj.correctNumberOfColumns = dataObj.correctNumberOfColumns ? dataObj.correctNumberOfColumns : ''
+
+  // use the first non-comment, non-blank, non-header row to determine correct number of columns
+  if (!line.startsWith('%') && line.trim().length > 0) {
+    dataObj.correctNumberOfColumns = dataObj.correctNumberOfColumns ? dataObj.correctNumberOfColumns : line.split(' ').length
+    if (dataObj.correctNumberOfColumns !== line.split(' ').length) {
+      dataObj.rowsWithWrongColumnNumbers.push(lineNum)
```

Comment:
Two suggestions:

1. Given whitespace is often visually ambiguous, use a variable to clarify.  (This obviously entails assigning it above, e.g. at top of this module.)
2. Assign `numColumns` once rather than evaluating it twice.

```suggestion
    const numColumns = line.split(tabCharacter).length
    dataObj.correctNumberOfColumns = dataObj.correctNumberOfColumns ? dataObj.correctNumberOfColumns : numColumns
    if (dataObj.correctNumberOfColumns !== numColumns) {
      dataObj.rowsWithWrongColumnNumbers.push(lineNum)
```